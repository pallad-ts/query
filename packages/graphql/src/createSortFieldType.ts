import {GraphQLEnumType} from "graphql";
import * as is from 'predicates'
import {camelCase} from "camel-case";

const assertSortableFields = is.assert(is.notEmptyArr);

/**
 * Creates enum for sortable fields
 */
export function createSortFieldType(baseName: string, sortableFields: readonly string[]) {
	assertSortableFields(sortableFields);
	return new GraphQLEnumType({
		name: `${baseName}_Sort_Field`,
		values: Object.fromEntries(
			sortableFields.map(x => {
				return [camelCase(x), {value: x}]
			})
		)
	})
}
