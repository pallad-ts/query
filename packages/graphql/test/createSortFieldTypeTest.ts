import {createSortFieldType} from "@src/createSortFieldType";

describe('createSortFieldType', () => {
	it('fails if sortable fields is empty array', () => {
		expect(() => {
			createSortFieldType('test', []);
		})
			.toThrowErrorMatchingSnapshot();
	});

	it.each([
		[['name', 'last_name']],
		[['first_name', 'id', 'createdAt', 'updatedAt']]
	])('creates enum field with camel cased values: %S', fields => {
		expect(
			createSortFieldType('EntityName', fields)
				.toConfig()
		)
			.toMatchSnapshot();
	});
})
