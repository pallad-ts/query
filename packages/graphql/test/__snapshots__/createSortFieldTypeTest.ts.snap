// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createSortFieldType creates enum field with camel cased values: %S 1`] = `
Object {
  "astNode": undefined,
  "description": undefined,
  "extensionASTNodes": Array [],
  "extensions": Object {},
  "name": "EntityName_Sort_Field",
  "values": Object {
    "lastName": Object {
      "astNode": undefined,
      "deprecationReason": undefined,
      "description": undefined,
      "extensions": Object {},
      "value": "last_name",
    },
    "name": Object {
      "astNode": undefined,
      "deprecationReason": undefined,
      "description": undefined,
      "extensions": Object {},
      "value": "name",
    },
  },
}
`;

exports[`createSortFieldType creates enum field with camel cased values: %S 2`] = `
Object {
  "astNode": undefined,
  "description": undefined,
  "extensionASTNodes": Array [],
  "extensions": Object {},
  "name": "EntityName_Sort_Field",
  "values": Object {
    "createdAt": Object {
      "astNode": undefined,
      "deprecationReason": undefined,
      "description": undefined,
      "extensions": Object {},
      "value": "createdAt",
    },
    "firstName": Object {
      "astNode": undefined,
      "deprecationReason": undefined,
      "description": undefined,
      "extensions": Object {},
      "value": "first_name",
    },
    "id": Object {
      "astNode": undefined,
      "deprecationReason": undefined,
      "description": undefined,
      "extensions": Object {},
      "value": "id",
    },
    "updatedAt": Object {
      "astNode": undefined,
      "deprecationReason": undefined,
      "description": undefined,
      "extensions": Object {},
      "value": "updatedAt",
    },
  },
}
`;

exports[`createSortFieldType fails if sortable fields is empty array 1`] = `"Assertion failed. Must be not an empty array"`;
